docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,497 @@
+
+#### `gardenadm token`
+
+After the first control plane node has been bootstrapped, it is very likely that we want to join additional nodes (either more control plane nodes to achieve high-availability, or worker nodes).
+As for regular shoot clusters, we will leverage the [`kubelet` TLS bootstrapping](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/) process.
+This means that a [bootstrap token](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/) is used by `kubelet` to request a client certificate via the `CertificateSigningRequest` API.
+
+We will provide commands to create and manage such bootstrap tokens:
+
+- `gardenadm token generate` will generate a random bootstrap token.
+- `gardenadm token create [token]` will create a bootstrap token.
+  If no `[token]` is provided, the command will generate a random token instead.
+- `gardenadm token delete [token-id]` will delete a bootstrap token.
+- `gardenadm token list` will list all bootstrap tokens on the server.
+
+> [!NOTE]
+> `kubeadm` goes one step further and implements a discovery process for the certificate authority of the control plane.
+> For this purpose, so-called "discovery tokens" are used.
+>
+> First, the CA certificate is retrieved via an insecure connection from a `ConfigMap` that got created as part of the `kubeadm init` process.
+> Next, the downloaded CA certificate is validated based on a hash of the public key that must be provided as part of `kubeadm join`.
+> Finally, as additional validation, the CA certificate is retrieved via secure connection and compared with the CA retrieved initially.
+>
+> We might implement a similar discovery mechanism for `gardenadm` as this project evolves.
+> However, we plan to first focus on the other, "more functional" aspects.
+
+#### `gardenadm join`
+
+Joining additional worker nodes is very similar to how it works for regular shoot clusters.
+This command instruments `gardener-node-agent` reconciliation logic to write the `systemd` units and needed files for the [`gardener-node-init`](../concepts/node-agent.md#installation-and-bootstrapping) component onto the disk of the machine.
+When this unit is started, it extracts the `gardener-node-agent` binary from its container image, and creates the `systemd` unit for it.
+`gardener-node-agent` will then use the provided bootstrap token to fetch the `OperatingSystemConfig` from a `Secret` residing in the `kube-system` namespace of the cluster.
+This `OperatingSystemConfig` contains the actual components like `kubelet`, `valitail`, etc.
+With the help of the same bootstrap token, `kubelet` will come up and register its `Node` resource, hence, successfully joins the cluster.
+
+The process for joining additional control plane nodes is very similar.
+The main difference is that the `OperatingSystemConfig` will contain the manifests of the static pods for the minimal standard Kubernetes control plane.
```

Comment:
How do additional control plane nodes retrieve private keys, e.g., front-proxy CA, client CA, etcd, ServiceAccount signing key?

Reply:
Just like the initial control plane node is bootstrapped, these secrets will probably be part of the `OperatingSystemConfig` itself as files.

Reply:
Are all static pods are started on each control plane node? 
Then, we could (temporary) reach states which are not ideal like an etcd cluster with two members or just inefficient like 5 control planes nodes with a  kube-controller-manager and kube-scheduler running on each node.

For the latter case, we could introduce `roles` for the control plane members, e.g. one per static pod.

Reply:
> Are all static pods are started on each control plane node?

Yes, all control plane nodes run a static KAPI,KCM,KSCHED, etc. pod.

> Then, we could (temporary) reach states which are not ideal like an etcd cluster with two members

Isn't this already the case with regular shoots upgraded to HA? Anyways, let's not talk about this now since the whole ETCD management is not clear yet.

> or just inefficient like 5 control planes nodes with a kube-controller-manager and kube-scheduler running on each node.

This depends on the rolling update strategy (in medium-touch case) and the `maxSurge` settings, but it shouldn't be a problem, or? This is a very temporary situation that will resolve within minutes (typically).

> For the latter case, we could introduce roles for the control plane members, e.g. one per static pod.

I didn't fully understand this proposal, what do you mean with roles?

Reply:
> > or just inefficient like 5 control planes nodes with a kube-controller-manager and kube-scheduler running on each node.
> 
> This depends on the rolling update strategy (in medium-touch case) and the `maxSurge` settings, but it shouldn't be a problem, or? This is a very temporary situation that will resolve within minutes (typically).

I don't get it completely 😅 
There won't be a deployment for the static pods, so there is no `maxSurge`. Thus, when you have 5 control plane node and every node is up and running, there will be always 5 instances of each static pod, won't there?

> 
> > For the latter case, we could introduce roles for the control plane members, e.g. one per static pod.
> 
> I didn't fully understand this proposal, what do you mean with roles?

Like you run `gardenadm join --role=kube-apiserver,kube-scheduler` and the machine joins the control-plane and runs `kube-apiserver` and `kube-scheduler` static pods only, but not `etcd` and `kube-controller-manager`.

Reply:
> There won't be a deployment for the static pods, so there is no `maxSurge`. Thus, when you have 5 control plane node and every node is up and running, there will be always 5 instances of each static pod, won't there?

Yes, sure, but the `MachineDeployment` has a `maxSurge` configuration as well. Hence, in the medium-touch case, when a rolling update is happening, this can control how many control plane nodes will be created in addition. In the high-touch scenario, we have to do in-place updates anyways.

> Like you run `gardenadm join --role=kube-apiserver,kube-scheduler` and the machine joins the control-plane and runs `kube-apiserver` and `kube-scheduler` static pods only, but not `etcd` and `kube-controller-manager`.

Can you explain a reasonable use-case for this? I haven't seen such setups in the past 7 years. All control plane nodes always run - you guessed it - the control plane components 😄. Yes, we have to anyway be careful with `etcd`, but I don't see why one would join a control plane node that just runs a subset of the components.

Reply:
I guess @oliver-goetz is referring to the inefficiency due to multiple pods running with leader election rendering all but one pod rather useless. For example, if we assume a control plane with 5 nodes only one `kube-scheduler` will actually work while the other 4 are just trying to become leader.

This obviously has room for improvement. However, I would argue that we need more experience to design a proper API for this. The proposed approach is a simple way how to get started. For obvious scenarios, e.g. single node control plane or 3-node control plane, it works as expected. If you scale your control plane in interesting ways, e.g. 5-node control plane, it may be less than optimal, but should still work.

If you agree with this we can add a section in the future work chapter.
WDYT?

Reply:
> > Like you run `gardenadm join --role=kube-apiserver,kube-scheduler` and the machine joins the control-plane and runs `kube-apiserver` and `kube-scheduler` static pods only, but not `etcd` and `kube-controller-manager`.
> 
> Can you explain a reasonable use-case for this? I haven't seen such setups in the past 7 years. All control plane nodes always run - you guessed it - the control plane components 😄. Yes, we have to anyway be careful with `etcd`, but I don't see why one would join a control plane node that just runs a subset of the components.

There is primarily one use-case, I have in mind which is scaling the control-plane. Imagine, there is an HA or non-HA etcd setup which is running fine, but the kube-apiserver is overloaded. It would be easier to just add another kube-apiserver node to the control-plane, without running an etcd here. Maybe this is just over engineering and we could just document "*only use an odd number of nodes*".
Deploying kube-controller-manager and kube-scheduler or not won't have a big impact.

Reply:
Added section in https://github.com/gardener/gardener/pull/10536/commits/99eefb454c4ebd6701eab79312bee3ddd8ae1f56.
PTAL.